test/system/src/test/java/io/pravega/test/system/ReadWriteAndScaleWithFailoverTest.java

```diff
@@ -0,0 +1,459 @@
+            }
+
+            //start writing asynchronoulsy with newly created writers
+            List<CompletableFuture<Void>> writerFutureList1 = writerList1.stream().map(writer ->
+                    startWriting(writer)).collect(Collectors.toList());
+
+            //run the failover test while scaling
+            performFailoverTest();
+
+            //wait for scaling
+            waitForScaling();
+
+            //bring the instances back to 3 before performing failover
+            controllerInstance.scaleService(3, true);
+            segmentStoreInstance.scaleService(3, true);
+            Thread.sleep(ZK_DEFAULT_SESSION_TIMEOUT);
+
+            //run the failover test after scaling
+            performFailoverTest();
+
+            log.info("Stop write flag status {}", testState.stopWriteFlag);
+            testState.stopWriteFlag.set(true);
+
+            log.info("Wait for writers execution to complete");
+            FutureHelpers.allOf(writerFutureList).get();
+
+            FutureHelpers.allOf(writerFutureList1).get();
+
+            log.info("Stop read flag status {}", testState.stopReadFlag);
+            testState.stopReadFlag.set(true);
+
+            log.info("Wait for readers execution to complete");
+            FutureHelpers.allOf(readerFutureList).get();
+
+            log.info("All writers have stopped. Setting stopReadFlag. Event Written Count:{}, Event Read " +
+                    "Count: {}", testState.eventWriteCount.get(), testState.eventsReadFromPravega.size());
+            assertEquals(testState.eventWriteCount.get(), testState.eventsReadFromPravega.size());
+            assertEquals(testState.eventWriteCount.get(), new TreeSet<>(testState.eventsReadFromPravega).size()); //check unique events.
+
+            log.info("Closing writers");
+            writerList.forEach(writer -> writer.close());
+            log.info("Closing readers");
+            readerList.forEach(reader -> reader.close());
+            log.info("Deleting readergroup {}", readerGroupName);
+            readerGroupManager.deleteReaderGroup(readerGroupName);
+        }
+        CompletableFuture<Boolean> sealStreamStatus = controller.sealStream(scope, STREAM_NAME);
+        log.info("Sealing stream {}", STREAM_NAME);
+        assertTrue(sealStreamStatus.get());
+        CompletableFuture<Boolean> deleteStreamStatus = controller.deleteStream(scope, STREAM_NAME);
+        log.info("Deleting stream {}", STREAM_NAME);
+        assertTrue(deleteStreamStatus.get());
+        CompletableFuture<Boolean> deleteScopeStatus = controller.deleteScope(scope);
+        log.info("Deleting scope {}", scope);
+        assertTrue(deleteScopeStatus.get());
+        log.info("Test {} succeeds ", "ReadWriteAndScaleWithFailover");
+
+        }
+
+        private void waitForScaling() throws InterruptedException, ExecutionException {
+
+            CompletableFuture<Void> testResult = Retry.withExpBackoff(10, 10, 40, ofSeconds(10).toMillis())
```

Comment:
This seems like it would be simpler, as a normal while loop. Backoff is not actually needed, and the exceptions are just used as flow control.